timemachine/ff/handlers/utils.py

```diff
@@ -67,3 +103,224 @@ def match_smirks(mol, smirks):
+    for bond in template_res_mol.GetBonds():
+        src_idx = bond.GetBeginAtomIdx()
+        dst_idx = bond.GetEndAtomIdx()
+        template_bonds[(src_idx, dst_idx)] = bond
+        template_bonds[(dst_idx, src_idx)] = bond
+
+    topology_res_mol.UpdatePropertyCache()
+    for bond in topology_res_mol.GetBonds():
+        src_idx = bond.GetBeginAtomIdx()
+        dst_idx = bond.GetEndAtomIdx()
+
+        template_src_idx = fwd_map[src_idx]
+        template_dst_idx = fwd_map[dst_idx]
+        k = (template_src_idx, template_dst_idx)
+        bond.SetBondType(template_bonds[k].GetBondType())
+        bond.SetIsAromatic(template_bonds[k].GetIsAromatic())
+
+    topology_res_mol.UpdatePropertyCache()
+
+
+def get_res_name(res_name: str) -> Tuple[str, bool, bool]:
+    """
+    Parameters
+    ----------
+    res_name:
+        Three letter of the residue. Can have an 'N' or 'C' prefix
+        to indicate a capped residue.
+
+    Return
+    ------
+    res_name: str
+        3 letter residue name
+    has_n_cap: bool
+        True if this residue is capped at the N terminus.
+    has_c_cap: bool
+        True if this residue is capped at the C terminus.
+    """
+    has_c_cap = False
+    has_n_cap = False
+    if len(res_name) == 4 and res_name[0] == "C":
+        res_name = res_name[1:]
+        has_c_cap = True
+    if len(res_name) == 4 and res_name[0] == "N":
+        has_n_cap = True
+        res_name = res_name[1:]
+
+    return res_name, has_n_cap, has_c_cap
+
+
+def add_n_cap(mol: Mol) -> Mol:
+    """
+    Add an N cap to the template residue mol.
+
+    NOTE: This only works properly for residues constructed
+    from the template, see `SMILES_BY_RES_NAME`.
+    """
+    mw = Chem.RWMol(mol)
+    mw.BeginBatchEdit()
+    # Need to adjust charge of the N to +1
+    query_mol = get_query_mol(Chem.MolFromSmiles(AMIDE_SMILES))
+    matches = mol.GetSubstructMatches(query_mol)
+
+    # find the N to add the cap
+    for i, atom in enumerate(mw.GetAtoms()):
+        if i not in matches[0]:
+            continue
+        if atom.GetSymbol() == "N":
+            atom.SetFormalCharge(+1)
+            atm_h = mw.AddAtom(Chem.Atom("H"))
+            mw.AddBond(atm_h, atom.GetIdx())
+            break
+    mw.CommitBatchEdit()
+    return mw
+
+
+def add_c_cap(mol: Mol) -> Mol:
+    """
+    Update the C-cap (COOH -> COO-).
+
+    NOTE: This only works properly for residues constructed
+    from the template, see `SMILES_BY_RES_NAME`.
+    """
+    mw = Chem.RWMol(mol)
+    mw.BeginBatchEdit()
+    # Need to adjust charge of the O to -1
+    query_mol = get_query_mol(Chem.MolFromSmiles(AMIDE_SMILES))
+    matches = mol.GetSubstructMatches(query_mol)
+
+    # find the O to adjust the charge
+    h_atom_idx = None
+    for i, atom in enumerate(mw.GetAtoms()):
+        if i not in matches[0]:
+            continue
+        if atom.GetSymbol() == "O":
+            if [bond.GetBondType() for bond in atom.GetBonds()] != [Chem.BondType.DOUBLE]:
```

Comment:
[non-blocking] I suppose this substructure matching logic could also be done with a smarts pattern, for a small maintainability improvement

(instead of matching the whole substructure implied by `AMIDE_SMILES="NCC(O)=O"`, then looping over in python to find the appropriate atom, could define an `AMIDE_SMARTS="[N]-[C]-[C](-[O:1])=O"` or similar, find the atom tagged 1... slightly less logic...)

Reply:
updated to simplify code in https://github.com/proteneer/timemachine/pull/1433/commits/7f9d9aaaf1e9c2644ecd36d8e600df897ecff22e